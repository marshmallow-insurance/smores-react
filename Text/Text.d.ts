import { FC, LabelHTMLAttributes, ReactNode } from 'react';
import { Color } from '../theme';
import { MarginProps } from '../utils/space';
/**
 * @deprecated Use only new Typo
 */
export type DeprecatedTypo = 'header-large' | 'header-medium' | 'header-small' | 'desc-heavy' | 'desc-medium' | 'desc-light' | 'desc-small' | 'base' | 'base-small' | 'base-xsmall' | 'label-large';
export type Typo = 'hero-alternate' | 'hero' | 'heading-alternate' | 'heading-large' | 'heading-medium' | 'heading-small' | 'headline-regular' | 'headline-small' | 'body-standfirst' | 'body-regular' | 'body-small' | 'caption' | 'label';
type Props = {
    children: ReactNode;
    tag?: any;
    className?: string;
    typo?: Typo | (string & Record<never, never>);
    align?: string;
    color?: Color;
    cursor?: string;
    title?: string;
} & MarginProps;
export type TextProps = Props & Omit<LabelHTMLAttributes<HTMLElement>, 'color'>;
export declare const Text: FC<TextProps>;
export {};
