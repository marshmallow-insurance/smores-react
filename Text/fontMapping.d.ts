import { Typo } from './Text';
export declare const fontStyleMapping: Record<Typo, string>;
