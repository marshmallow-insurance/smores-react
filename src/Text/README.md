# < Text />

## Props:

```ts
tag:        any;    /* HTML tag    */
className?: string;
typo?:      Typo | (string & Record<never, never>); /* typography  */
align?:     string; /* text-align  */
color?:     Color; /* from theme */
cursor?:    string;
```

## Usage

```js
import {Text} from '@mrshmllw/smores-react';

const App = () => (
  <Text tag='h2' typo='hero-alternate' color='secondary'>Hey you</Text>
);
```

## Typography:

| name | size mobile (px) | size desktop (px) | weight |
|------|:-----------:|:------------:|-------------|
| hero-alternate | 40 | 56 | bold |
| hero | 32 | 40 | medium |
| heading-large | 28 | 32 | medium |
| heading-medium | 24 | 24 | medium |
| heading-small | 20 | 20 | medium |
| headline-regular | 16 | 16 | medium |
| headline-small | 14 | 14 | medium |
| **body-regular** | **16** | **16** | **normal** |
| body-small | 14 | 14 | normal |
| caption | 12 | 12 | normal |
| label | 10 | 10 | medium |
