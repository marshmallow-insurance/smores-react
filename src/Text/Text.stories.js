import React from 'react'
import { Text } from './Text'

export default {
  title: 'Text',
  component: Text,
}

const Template = (args) => (
  <Text {...args}>The quick brown fox jumps over the lazy dog</Text>
)

export const Default = Template.bind({})

Default.args = {
  tag: 'p',
}

export const Label = Template.bind({})

Label.args = {
  tag: 'label',
  typo: 'label',
  color: 'secondary',
}

export const WithTitle = Template.bind({})

WithTitle.args = {
  tag: 'p',
  title: 'a11y title tooltip',
}
