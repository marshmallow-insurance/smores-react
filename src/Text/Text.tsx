import React, { FC, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { theme } from '../theme'

interface IText {
  /** typography class name to apply predefined styles */
  typo: string
  /** HTML tag  */
  as: any
  /** text-align  */
  align: string
  /** color from the theme  */
  color: string
  cursor: string
}

type Props = {
  children: ReactNode
  tag: any
  className?: string
  typo?: string
  align?: string
  color?: string
  cursor?: string
  title?: string
}

export const Text: FC<Props> = ({
  children,
  typo,
  className = '',
  tag = 'p',
  align = 'left',
  color = 'coreSecondary',
  cursor = 'inherit',
  title = '',
}) => (
  <Container
    as={tag}
    className={className}
    typo={typo || 'base'}
    align={align}
    color={color}
    cursor={cursor}
    title={title}
  >
    {children}
  </Container>
)

const Container = styled.p<IText>(
  ({ align, color, cursor, typo }) => css`
    /** PREDEFINED TYPOGRAPHY STYLES */

    /* Header Large */
    ${typo === 'header-large' &&
    css`
      font-size: 24px;
      line-height: 31px;
      font-weight: 500;

      @media (min-width: 768px) {
        font-size: 42px;
        line-height: 54px;
      }
    `}

    /* Header medium */
  ${typo === 'header-medium' &&
    css`
      font-size: 21px;
      line-height: 27px;
      font-weight: 500;

      @media (min-width: 768px) {
        font-size: 32px;
        line-height: 41px;
      }
    `}

  /* Header small */
  ${typo === 'header-small' &&
    css`
      font-size: 18px;
      line-height: 23px;
      font-weight: 500;

      @media (min-width: 768px) {
        font-size: 21px;
        line-height: 27px;
      }
    `}

  /* Descriptor Heavy */
  ${typo === 'desc-heavy' &&
    css`
      font-size: 14px;
      line-height: 20px;
      font-weight: 700;

      @media (min-width: 768px) {
        font-size: 16px;
        line-height: 23px;
      }
    `}

  /* Descriptor Medium */
  ${typo === 'desc-medium' &&
    css`
      font-size: 14px;
      line-height: 20px;
      font-weight: 500;

      @media (min-width: 768px) {
        font-size: 16px;
        line-height: 23px;
      }
    `}

  /* Descriptor Light */
  ${typo === 'desc-light' &&
    css`
      font-size: 14px;
      line-height: 20px;
      font-weight: 400;

      @media (min-width: 768px) {
        font-size: 16px;
        line-height: 23px;
      }
    `}

  /* Descriptor Small */
  ${typo === 'desc-small' &&
    css`
      font-size: 12px;
      line-height: 16px;
      font-weight: 400;

      @media (min-width: 768px) {
        font-size: 14px;
        line-height: 18px;
      }
    `}

  /* Base */
  ${typo === 'base' &&
    css`
      font-size: 14px;
      line-height: 23px;
      font-weight: 400;

      @media (min-width: 768px) {
        font-size: 16px;
        line-height: 26px;
      }
    `}

  /* Base Small */
  ${typo === 'base-small' &&
    css`
      font-size: 12px;
      line-height: 20px;
      font-weight: 400;

      @media (min-width: 768px) {
        font-size: 14px;
        line-height: 23px;
      }
    `}

  /* Base Extra Small */
  ${typo === 'base-xsmall' &&
    css`
      font-size: 10px;
      line-height: 16px;
      font-weight: 400;

      @media (min-width: 768px) {
        font-size: 12px;
        line-height: 19px;
      }
    `}

  /* Label */
  ${typo === 'label' &&
    css`
      font-size: 8px;
      line-height: 100%;
      font-weight: 700;
      text-transform: uppercase;
      letter-spacing: 0.75px;

      @media (min-width: 768px) {
        font-size: 10px;
      }
    `}
  
  /* Label Large */
  ${typo === 'label-large' &&
    css`
      font-size: 12px;
      line-height: 100%;
      font-weight: 700;
      text-transform: uppercase;
      letter-spacing: 0.75px;
    `}

  margin: 0;
    padding: 0;
    color: ${theme.colors[color]};
    text-align: ${align};
    cursor: ${cursor};
    letter-spacing: 0.15px;
  `,
)
