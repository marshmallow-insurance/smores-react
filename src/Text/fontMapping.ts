import { theme } from '../theme'
import { Typo } from './Text'

export const fontStyleMapping: Record<Typo, string> = {
  'hero-alternate': `
    font-size: 40px;
    font-family: 'ClearfaceITC';
    font-weight: ${theme.font.weight.bold};
    line-height: 44px;

    @media (min-width: 768px) {
      font-size: 56px;
      line-height: 64px;
    }
  `,
  hero: `
    font-size: 32px;
    font-weight: ${theme.font.weight.medium};
    line-height: 40px;

    @media (min-width: 768px) {
      font-size: 40px;
      line-height: 44px;
    }
  `,
  'heading-large': `
    font-size: 28px;
    font-weight: ${theme.font.weight.medium};
    line-height: 32px;

    @media (min-width: 768px) {
      font-size: 32px;
      line-height: 40px;
    }
  `,
  'heading-medium': `
    font-size: 24px;
    font-weight: ${theme.font.weight.medium};
    line-height: 32px;
  `,
  'heading-small': `
    font-size: 20px;
    font-weight: ${theme.font.weight.medium};
    line-height: 28px;
  `,
  'headline-regular': `
    font-size: 16px;
    font-weight: ${theme.font.weight.medium};
    line-height: 20px;
  `,
  'headline-small': `
    font-size: 14px;
    font-weight: ${theme.font.weight.medium};
    line-height: 20px;
  `,
  'body-standfirst': `
    font-size: 18px;
    font-weight: ${theme.font.weight.normal};
    line-height: 24px;
  `,
  'body-regular': `
    font-size: 16px;
    font-weight: ${theme.font.weight.normal};
    line-height: 20px;
  `,
  'body-small': `
    font-size: 14px;
    font-weight: ${theme.font.weight.normal};
    line-height: 20px;
  `,
  caption: `
    font-size: 12px;
    font-weight: ${theme.font.weight.normal};
    line-height: 16px;
  `,
  label: `
    font-size: 10px;
    line-height: 12px;
    font-weight: ${theme.font.weight.medium};
    text-transform: uppercase;
  `,
}
