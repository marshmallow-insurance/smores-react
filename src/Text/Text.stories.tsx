import React from 'react'
import styled from 'styled-components'
import { Text, TextProps } from './Text'
import { Box } from '../Box'
import { Typo } from './Text'
import { fontStyleMapping } from './fontMapping'

export default {
  title: 'Text',
  component: Text,
}

const typos = Object.keys(fontStyleMapping)

const deprecatedTypos = [
  'header-large',
  'header-medium',
  'header-small',
  'desc-heavy',
  'desc-medium',
  'desc-light',
  'desc-small',
  'base',
  'base-small',
  'base-xsmall',
]

const Template = (props: TextProps) => (
  <Text {...props}>The quick brown fox jumps over the lazy dog</Text>
)

const TypoCollection = ({
  typos,
}: {
  typos: Readonly<Typo[]> | Readonly<string[]>
}) => {
  return (
    <Box>
      <Grid>
        <Text tag="p" typo="base">
          Name
        </Text>
        <Text tag="p" typo="base">
          Single Line
        </Text>
        <Text tag="p" typo="base">
          Paragraph
        </Text>
      </Grid>
      {typos.map((typo) => (
        <Grid key={typo}>
          <Text tag="p" typo="base" color="sesame">
            {typo}
          </Text>
          <Text tag="p" typo={typo} color="liquorice">
            They waited patiently for what seemed a very long time.
          </Text>
          {!['hero-alternate', 'hero', 'label'].includes(typo) && (
            <Text tag="p" typo={typo} color="liquorice">
              They waited patiently for what seemed a very long time. They
              waited patiently for what seemed a very long time.
            </Text>
          )}
        </Grid>
      ))}
    </Box>
  )
}

const Grid = styled(Box)`
  display: grid;
  grid-template-columns: 0.4fr 1.2fr 1fr;
  margin-bottom: 24px;
  gap: 10px;
`

const TypographyTemplate = () => <TypoCollection typos={typos} />

const DeprecatedTemplate = () => <TypoCollection typos={deprecatedTypos} />

export const Default = Template.bind({})

Default.args = {
  tag: 'p',
}

export const Label = Template.bind({})

Label.args = {
  tag: 'label',
  typo: 'label',
  color: 'liquorice',
}

export const WithTitle = Template.bind({})

WithTitle.args = {
  tag: 'p',
  title: 'a11y title tooltip',
}

export const Collection = TypographyTemplate.bind({})

export const DeprecatedTypos = DeprecatedTemplate.bind({})
