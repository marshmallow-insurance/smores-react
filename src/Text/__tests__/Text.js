import React from 'react';
import {render} from '@testing-library/react';
import 'jest-styled-components';

import {Text} from '../Text';

test('renders', () => {
  const {container} = render(
    <Text>The quick brown fox jumps over the lazy dog</Text>,
  );
  expect(container.firstChild).toMatchSnapshot();
});
