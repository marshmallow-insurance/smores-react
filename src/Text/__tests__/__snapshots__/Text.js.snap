// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 20px;
  text-align: left;
  cursor: inherit;
  color: #292924;
}

<p
  class="c0"
  color="liquorice"
  cursor="inherit"
  title=""
>
  The quick brown fox jumps over the lazy dog
</p>
`;
